morebits.js

```diff
@@ -4566,215 +4577,1725 @@ Morebits.wiki.page = function(pageName, status) {
+		notifyBots: false,
+		notifyIndef: false,
+		notifySelf: false,
+		notifySkipTemplates: null,
+		notifySkipLink: null,
+		pageobjectFunctions: null,
+
+		// Internals
+		// In theory mw.user.tokens is available as a fallback,
+		// but since we're always loading there's really no need
+		csrfToken: null,
+		userrightsToken: null,
+		userApi: null,
+		userLoaded: false,
+		blockApi: null,
+		unblockApi: null,
+		groupsApi: null,
+		actionResponse: null,
+
+		// Callbacks
+		onLoadSuccess: null,
+		onLoadFailure: null,
+		onBlockSuccess: null,
+		onBlockFailure: null,
+		onUnblockSuccess: null,
+		onUnblockFailure: null,
+		onGroupsSuccess: null,
+		onGroupsFailure: null,
+		onNotifySuccess: null,
+		onNotifyFailure: null
+	};
 
-		var query = {
-			action: 'parse',
-			prop: 'text',
-			pst: 'true',  // PST = pre-save transform; this makes substitution work properly
-			text: wikitext,
-			title: pageTitle || mw.config.get('wgPageName'),
-			disablelimitreport: true,
+	var emptyFunction = function() { };
+
+	/**
+	 * Loads info about the user.  Required before (nearly) all of the
+	 * object methods, but will be done automatically if forgotten.  Note
+	 * that unlike {@link Morebits.wiki.page#load}, the `onSuccess` callback
+	 * is not required.
+	 *
+	 * @param {Function} [onSuccess] - Callback function which is called when the load has succeeded.
+	 * @param {Function} [onFailure] - Callback function which is called when the load fails.
+	 */
+	this.load = function(onSuccess, onFailure) {
+		ctx.onLoadSuccess = onSuccess;
+		ctx.onLoadFailure = onFailure || emptyFunction;
+
+		ctx.loadQuery = {
+			action: 'query',
+			// Potential expansions:
+			// list=usercontribs to get timestamp of user's last
+			// edit, can skip if haven't edited in X days
+			// list=allusers for global attached stats
+			// list=globalblocks
+			// meta=globaluserinfo for locked status
+			list: 'blocks|users|logevents',
+
+			// groups technically redundant to implicit+groupmemberships, but meh
+			usprop: 'registration|editcount|rights|groups|implicitgroups|groupmemberships',
+			ususers: ctx.userName,
+
+			// bkusers or bkip set below as appropriate
+			bkprop: 'id|user|by|timestamp|expiry|reason|flags|restrictions|range',
+
+			// Just to know if there is a block log.  Logs users, IPs, and CIDR blocks,
+			// but note: no entries present for an IP caught within a range block.
+			// Moreover, semi-busted on ranges, see [[phab:T270737]] and [[phab:T146628]].
+			// Basically, logevents doesn't treat functionally-equivalent ranges
+			// as equivalent, meaning functionally-equivalent IP ranges may be
+			// misinterpreted.  Without logevents redirecting (like Special:Block does)
+			// we would need a function to parse ranges, which is a pain.
+			// IPUtils has code, but it'd be a lot of cruft for one purpose.
+			letype: 'block',
+			letitle: userTitle.toText(),
+			lelimit: 1,
+
+			// Get the talk page content, categories, etc.
+			titles: ctx.talkTitle,
```

Comment:
Originally I had a `setLoadTalk` method to optionally load the talk, but since we rely on it for optouts, it was just getting in the way and more trouble than it was worth.